File: backend/src/utils/calculator.ts
export type ProductParams = {
  weightPerPiece: number; // WghPcs
  quantityPerMeasurement: number; // Qty
  tolerance: number; // DevW
};

export function standardWeight({ weightPerPiece, quantityPerMeasurement }: ProductParams): number {
  return weightPerPiece * quantityPerMeasurement;
}

export function minWeight({ weightPerPiece, quantityPerMeasurement }: ProductParams): number {
  return standardWeight({ weightPerPiece, quantityPerMeasurement, tolerance: 0 }) - weightPerPiece / 2;
}

export function maxWeight({ weightPerPiece, quantityPerMeasurement }: ProductParams): number {
  return standardWeight({ weightPerPiece, quantityPerMeasurement, tolerance: 0 }) + weightPerPiece / 2;
}

export function dMin({ weightPerPiece, quantityPerMeasurement, tolerance }: ProductParams): number {
  return standardWeight({ weightPerPiece, quantityPerMeasurement, tolerance }) - tolerance;
}

export function dMax({ weightPerPiece, quantityPerMeasurement, tolerance }: ProductParams): number {
  return standardWeight({ weightPerPiece, quantityPerMeasurement, tolerance }) + tolerance;
}

export type Classification = 'GREEN' | 'YELLOW' | 'RED';

export function classify(weight: number, p: ProductParams): Classification {
  const s = standardWeight(p);
  const mn = s - p.weightPerPiece / 2;
  const mx = s + p.weightPerPiece / 2;
  const dmn = s - p.tolerance;
  const dmx = s + p.tolerance;

  if (weight >= dmn && weight <= dmx) return 'GREEN';
  if ((weight < dmn && weight >= mn) || (weight > dmx && weight <= mx)) return 'YELLOW';
  return 'RED';
}

File: backend-spring/src/main/java/com/example/eikensystem/web/Calculator.java
package com.example.eikensystem.web;

import org.springframework.stereotype.Component;

@Component
public class Calculator {
    // qty = quantityPerMeasurement
    public double standardWeight(double weightPerPiece, int qty) {
        return weightPerPiece * qty;
    }
    // Min = Std - wpp/2
    public double minWeight(double weightPerPiece, int qty) {
        return standardWeight(weightPerPiece, qty) - weightPerPiece / 2.0;
    }
    // Max = Std + wpp/2
    public double maxWeight(double weightPerPiece, int qty) {
        return standardWeight(weightPerPiece, qty) + weightPerPiece / 2.0;
    }
    // DMin = Std - wpp/4
    public double dMin(double std, double weightPerPiece) {
        return std - weightPerPiece / 4.0;
    }
    // DMax = Std + wpp/4
    public double dMax(double std, double weightPerPiece) {
        return std + weightPerPiece / 4.0;
    }
    // qty = quantityPerMeasurement; tolerance param kept for API compat but devW is always wpp/4
    public String classify(double weight, double weightPerPiece, int qty, double tolerance) {
        double s   = standardWeight(weightPerPiece, qty);
        double mn  = s - weightPerPiece / 2.0;
        double mx  = s + weightPerPiece / 2.0;
        double dmn = s - weightPerPiece / 4.0;
        double dmx = s + weightPerPiece / 4.0;
        if (weight >= dmn && weight <= dmx) return "GREEN";
        if ((weight < dmn && weight >= mn) || (weight > dmx && weight <= mx)) return "YELLOW";
        return "RED";
    }
}
